in-progress/5040-native-merkle-trees-napi.md

```diff
@@ -0,0 +1,341 @@
+    return deferred.Promise();
+  }
+
+  bool include_uncomitted = info[2].As<Napi::Boolean>();
+
+  // pointer to helper class for async code (gets cleaned up later), see below
+  auto* tree_op = new bb::world_state::TreeOp(env, deferred, [=]() {
+    bb::crypto::merkle_tree::Signal signal(1);
+    bb::fr leaf(0);
+    auto callback = [&](bb::fr& value) {
+      leaf = value;
+      signal.signal_level(0);
+    };
+    _world_state_svc[tree_id].get_leaf_value(leaf_index, include_uncomitted);
+    signal.wait_for_level(0);
+    return leaf;
+  });
+
+  tree_op ->Queue();
+
+  return deferred.Promise();
+}
+// etc.
+
+// init the module
+Napi::Function WorldStateAddon::get_class(Napi::Env env)
+{
+    return DefineClass(env, "WorldState",
+      {
+          WorldStateAddon::InstanceMethod("getLeafValue", &WorldStateAddon::getLeafValue),
+          // other instance methods
+      });
+}
+
+Napi::Object Init(Napi::Env env, Napi::Object exports)
+{
+    Napi::String name = Napi::String::New(env, "WorldState");
+    exports.Set(name, WorldStateAddon::get_class(env));
+    return exports;
+}
+
+NODE_API_MODULE(addon, Init)
+```
+
+> [!NOTE]
+> The instance methods on the C++ class will be exported as instance methods on the JavaScript instance too.
+> Instance methods _must_ return a Napi::Value (ie. `any` in TS-land), even though the method returns something more specific (e.g. a Promise) and accept a single Napi::Callback parameter.
+> Instance methods can not be `const`.
+
+### Classes & instances
+
+Exported classes from the C++ side can be used an instantiated from NodeJS. Node Addon API is responsible for the glue code that ties to two together (ie. calling a function on the JS object calls the appropriate function in C++). The JS instance is a reference to the instance inside C++. The C++ instance is able to refer instantiate any other classes or allocate and access as much memory as needed.
+
+When an instance is garbage collected on the TS-side, the destructor is called on the C++ side.
+
+### Passing data between NodeJS and C++
+
+The `Napi` namespace on the C++ contains helper classes to deal with JS primitive values. Strings, numbers, bigints, buffers, arrays, typed arrays and even functions can be freely passed between the two environments.
+
+More complex data structures must be serialized/deserialized. We will msgpack for this as it's already implemented in the C++ code
+
+[`Napi::Value` documentation](https://github.com/nodejs/node-addon-api/blob/cc06369aa4dd29e585600b8b47839c1297df962d/doc/value.md)
+
+### Message passing
+
+Instead of exporting a functions for each operation on the world state, we could instead leverage the existing message passing interface, only instead sending the message across the TS/C++ boundary. This would simplify the module initialization code on the C++ side (only requiring we export a single function) and we'd benefit from easily serializing data types with msgpack.
```

Comment:
If we implemented this, does it make sense to keep maintaining the message passing interface? Feels like it'd be simpler to deprecate that one, and just have one exported function for each actual tree function, right? No strong feelings here though.

Reply:
I think it'll come down to the kind of async ops that we'll run inside `AsyncWorker` instances and how we design the code (someone with more C++ experience could chime in here). It might be easier to refactor the generic `tree_op_callback` to return a msgpacked value (currently in this spec the callback must return a `bb::fr`) as then it becomes trivial to send it back to JS as a buffer vs serializing to POJOs in the `OnOk` callback.

Reply:
If we kept the message passing interface, then wouldn't we not need to worry about keeping types in parity across TS and cpp? And we wouldn't need to deserialize `any` in TS? And minor benefit that other clients of bb (if/when that day should come) that want to use these trees without the nodeapi could more easily integrate them?